### PR TITLE
Fix Codable vtable layout

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes <rdar://problem/35647420> and <https://bugs.swift.org/browse/SR-6468>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@lorentey This introduces a mechanism for derived conformances to define class methods with deterministic vtable order. Just remember to update `TypeChecker::finalizeType()` to force the conformance, and remember to call `setSynthesized()` on every method you add to the class.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: include/swift/SIL/SILVTableVisitor.h [hunk 1]
```diff
@@ -18,11 +18,55 @@
...
+                return lhs.first < rhs.first;
+              });
+  }
+
+  decltype(elts)::const_iterator begin() const {
+    assert(sorted);
+    return elts.begin();
+  }
+
+  decltype(elts)::const_iterator end() const {
```
Comment (DougGregor) [inline include/swift/SIL/SILVTableVisitor.h]: I wonder if some of the less-capable C++ compilers we build with will handle this `decltype` idiom. I guess we'll find out

File: include/swift/SIL/SILVTableVisitor.h [hunk 2]
```diff
@@ -76,19 +120,47 @@ template <class T> class SILVTableVisitor {
...
+
+      maybeAddMember(member);
+    }
+
+    if (synthesizedMembers.empty())
+      return;
+
+    synthesizedMembers.sort();
+
+    for (auto pair : synthesizedMembers) {
```
Comment (DougGregor) [inline include/swift/SIL/SILVTableVisitor.h]: `const auto &pair` to avoid the extraneous copies, please

File: lib/Sema/TypeCheckDecl.cpp [hunk 3]
```diff
@@ -8005,40 +8013,42 @@ static void finalizeType(TypeChecker &TC, NominalTypeDecl *nominal) {
...
+    // We need to add implicit initializers and dtors because it
+    // affects vtable layout.
     TC.addImplicitConstructors(CD);
     CD->addImplicitDestructor();
+
+    // We need the superclass vtable layout as well.
+    TC.requestSuperclassLayout(CD);
+
+    auto useConformance = [&](ProtocolDecl *protocol) {
+      (void)TC.conformsToProtocol(
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: This is going to check the conformance even if it's written on an extension, which will perform some unnecessary work. I guess that's going to be fairly rare for a class, but if you're concerned about it, you could drop the `Used` flag here and only force the completion of the conformance when it's `DeclContext` is the nominal type itself.

File: lib/Sema/TypeCheckDecl.cpp [hunk 4]
```diff
@@ -9012,48 +9024,20 @@ void TypeChecker::synthesizeMemberForLookup(NominalTypeDecl *target,
...
+  // If the target conforms to either and the conformance has not yet been
+  // evaluated, then we should do that here.
+  //
+  // Try to synthesize Decodable first. If that fails, try to synthesize
+  // Encodable. If either succeeds and CodingKeys should have been
+  // synthesized, it will be synthesized.
+  auto *decodableProto = Context.getProtocol(KnownProtocolKind::Decodable);
+  auto *encodableProto = Context.getProtocol(KnownProtocolKind::Encodable);
+  if (!evaluateTargetConformanceTo(decodableProto))
+    (void)evaluateTargetConformanceTo(encodableProto);
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: You don't want to check for the name `CodingKeys` at all?

**Review by DougGregor — APPROVED** [review]

File: lib/Sema/TypeCheckDecl.cpp (hunk 4 above)
Comment (slavapestov) [reply lib/Sema/TypeCheckDecl.cpp]: It's checked with an early return at the top of the method

File: include/swift/SIL/SILVTableVisitor.h (hunk 2 above)
Comment (slavapestov) [reply include/swift/SIL/SILVTableVisitor.h]: Fixed

File: lib/Sema/TypeCheckDecl.cpp (hunk 3 above)
Comment (slavapestov) [reply lib/Sema/TypeCheckDecl.cpp]: Good point, fixed

**Comment by lorentey** [comment]
@slavapestov Thank you, this looks great! 🎉

**Comment by itaiferber** [comment]
This looks good too! A lot of the changes here were also made in #16054 too; should this subsume that PR?

(Also, both PRs appear to have the same test failures in `init` checking from ObjC; possibly related to the changes in the `DeclChecker`?)

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**